[[ Bug 23096 ]] Implement translated environment AppleScript workaround

montegoulding opened the pull request:
This patch implements a workaround for executing AppleScript from an engine
translated by Rosetta by executing the script via `osascript` executing
natively.

runrevmark commented on engine/src/dskmac.cpp:
```diff
@@ -4627,9 +4627,63 @@ struct MCMacDesktop: public MCSystemInterface, public MCMacSystemService
+
     virtual void DoAlternateLanguage(MCStringRef p_script, MCStringRef p_language)
     {
+		uinteger_t t_environment_translated = 0;
```
This code is only needed on Big Sur and above (use MCmajorosversion).

runrevmark commented on engine/src/dskmac.cpp:
```diff
@@ -4627,9 +4627,63 @@ struct MCMacDesktop: public MCSystemInterface, public MCMacSystemService
+		uinteger_t t_environment_translated = 0;
+		if (MCStringIsEqualToCString(p_language, "AppleScript", kMCStringOptionCompareCaseless))
+		{
+			MCresult->clear();
```
We don't need to use shell to determine whether we are running under a translated environment we can easily do that in C - https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment:
```
int processIsTranslated() {
   int ret = 0;
   size_t size = sizeof(ret);
   if (sysctlbyname("sysctl.proc_translated", &ret, &size, NULL, 0) == -1) 
   {
      if (errno == ENOENT)
         return 0;
      return -1;
   }
   return ret;
}
```
Also we can cache the result of testing for rosetta in a local static as it isn't going to change for the lifetime of the process :)

runrevmark commented on docs/notes/bugfix-23096.md:
```diff
@@ -0,0 +1 @@
+# Implement workaround for AppleScript execution in Rosetta tranalated environment
```
Bug fix notes should describe what has been fixed - in this case I believe it fixes a significant delay when using AppleScript to target some Applications when running on M1 architecture macs?

runrevmark commented on engine/src/dskmac.cpp:
```diff
@@ -4627,9 +4628,75 @@ struct MCMacDesktop: public MCSystemInterface, public MCMacSystemService
+
     virtual void DoAlternateLanguage(MCStringRef p_script, MCStringRef p_language)
     {
+		if (MCmajorosversion >= MCOSVersionMake(10,16,0)
```
breaks should go after operators

runrevmark commented on engine/src/dskmac.cpp:
```diff
@@ -4627,9 +4628,75 @@ struct MCMacDesktop: public MCSystemInterface, public MCMacSystemService
+		"return \"{\" & tValue & \"}\";" \
+	"end switch"
+
+	int process_is_translanslated()
```
I suggest
```
static bool GetProcessIsTranslated(void)
{
    static int s_state = -1;
    if (s_state == -1)
    {
        ....
    }
    return s_state == 1;
}
```

The ordering of the stuff in this file is a little bit messy, but the definition shouldn't be in the class, maybe put it at the top where you currently define the file-local static.

(Rationale for putting the static in the function is that it encapsulates it better).

montegoulding committed:
[[ Bug 23096 ]] Implement translated environment AppleScript workaround

This patch implements a workaround for executing AppleScript from an engine
translated by Rosetta by executing the script via `osascript` executing
natively.

runrevmark commented:
@livecode-vulcan review ok 6e123cc

livecode-vulcan commented:
💙 review by @runrevmark ok 6e123cc39512db5b2e964b428de940199c1dccad

livecode-vulcan commented:
😎 test success 6e123cc39512db5b2e964b428de940199c1dccad

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/985)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/393)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/392)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/391)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/392)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2510)
* try-community-universal-ios-iphoneos14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.4/builds/39)
* try-community-universal-ios-iphoneos14.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.1/builds/88)
* try-community-universal-ios-iphonesimulator14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.4/builds/39)
* try-community-universal-ios-iphonesimulator14.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.1/builds/86)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2537)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2526)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2524)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2561)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2497)
